kevm-pyk/src/kevm_pyk/foundry.py

```diff
@@ -599,6 +603,101 @@ def _short_info(cterm: CTerm) -> Iterable[str]:
     return '\n'.join(res_lines)
 
 
+def foundry_koverage(foundry_root: Path, contracts: Iterable[str]) -> None:
+    foundry = Foundry(foundry_root)
+    apr_proofs_dir = foundry.out / 'apr_proofs'
+    artifacts = foundry.contracts
+    contracts = set(contracts)
+    len_contracts = len(contracts)
+
+    call_cells: dict[bytes, list[tuple[KInner, KCFG.Node, list[KCFG.Node]]]] = {}
+
+    for name, contract in artifacts.items():
+        if (len_contracts != 0 and name in contracts) or len_contracts == 0:
+            tests = contract.tests
+            for test in tests:
+                proof_digest = foundry.proof_digest(name, test)
+                if APRProof.proof_exists(proof_digest, apr_proofs_dir):
+                    apr_proof = APRProof.read_proof(proof_digest, apr_proofs_dir)
+                    if apr_proof.status is ProofStatus.PASSED or True:
+                        kcfg = apr_proof.kcfg
+                        nodes = kcfg.nodes
+                        leaves = kcfg.leaves
+
+                        if len(nodes) > 1:
+                            base_node = nodes[1]
+
+                            for node in nodes:
+                                # print(node.cterm.constraints)
+                                cterm = node.cterm
+                                kast = cterm.cell('K_CELL')
+                                if type(kast) is KSequence:
+                                    for item in kast.items:
+                                        if type(item) is KApply and item.label.name.startswith('#callWithCode'):
+                                            # print(node)
+                                            call_args = item.args
+                                            # TODO: parse this and store the caller
+                                            # next, we should try to get the origin and other interesting values
+                                            call_args[0]
+                                            code = call_args[3]
+                                            if type(code) is KToken:
+                                                acct_code = code.token
+                                                acct_hash = hashlib.sha3_256(bytes(acct_code, 'UTF-8')).digest()
+
+                                                val = call_cells.get(acct_hash)
+                                                if val is None:
+                                                    val = []
+
+                                                val.append((cterm.config, base_node, leaves))
+                                                call_cells[acct_hash] = val
+
+                                                break
```

Comment:
We can use pyk's pattern matching here instead, something like:

```
k_cell_pattern = KSequence([KApply('#call_with_code_klabel', [KVariable('_A1'), KVariable('_A2'), KVariable('CODE'), ...])])

k_match = k_cell_pattern.match(cterm.cell('K_CELL'))
if k_match is not None:
    acct_code = k_match['CODE']
```

Reply:
Done